[lldb][Mach-O] Don't read symbol table of specially marked binary

jasonmolenda opened the pull request:
We have a binary image on Darwin that has no code, only metadata. It has a large symbol table with many external symbol names that will not be needed in the debugger.  And it is possible to not have this binary on the debugger system - so lldb must read all of the symbol names out of memory, one at a time, which can be quite slow.

We're adding a section __TEXT,__lldb_no_nlist, to this binary to indicate that lldb should not read the nlist symbols for it when we are reading out of memory.  If lldb is run with an on-disk version of the binary, we will load the symbol table as we normally would, there's no benefit to handling this binary differently.

I added a test where I create a dylib with this specially named section, launch the process.  The main binary deletes the dylib from the disk so lldb is forced to read it out of memory.  lldb attaches to the binary, confirms that the dylib is present in the process and is a memory Module.  If the binary is not present, or lldb found the on-disk copy because it hasn't been deleted yet, we delete the target, flush the Debugger's module cache, sleep and retry, up to ten times.  I create the specially named section by compiling an assembly file that puts a byte in the section which makes for a bit of a messy Makefile (the pre-canned actions to build a dylib don't quite handle this case) but I don't think it's much of a problem.  This is a purely skipUnlessDarwin test case.

rdar://146167816

jasonmolenda committed:
[lldb][Mach-O] Don't read symbol table of specially marked binary

We have a binary image on Darwin that has no code, only metadata.
It has a large symbol table with many external symbol names that
will not be needed in the debugger.  And it is possible to not have
this binary on the debugger system - so lldb must read all of the
symbol names out of memory, one at a time, which can be quite slow.

We're adding a section __TEXT,__lldb_no_nlist, to this binary to
indicate that lldb should not read the nlist symbols for it when
we are reading out of memory.  If lldb is run with an on-disk version
of the binary, we will load the symbol table as we normally would,
there's no benefit to handling this binary differently.

I added a test where I create a dylib with this specially named
section, launch the process.  The main binary deletes the dylib
from the disk so lldb is forced to read it out of memory.  lldb
attaches to the binary, confirms that the dylib is present in
the process and is a memory Module.  If the binary is not present,
or lldb found the on-disk copy because it hasn't been deleted yet,
we delete the target, flush the Debugger's module cache, sleep and
retry, up to ten times.  I create the specially named section by
compiling an assembly file that puts a byte in the section which
makes for a bit of a messy Makefile (the pre-canned actions to
build a dylib don't quite handle this case) but I don't think it's
much of a problem.  This is a purely skipUnlessDarwin test case.

rdar://146167816

jasonmolenda commented:
The change to ObjectFileMachO looks a little larger than it really is because i moved the SectionSP initializations ~100 lines earlier in ParseSymtab than they were.  I'm scanning for this new section the same way we scan for eh_frame.  ObjectFileMachO already had "load level" overrides for reading ObjectFiles out of memory, so it was a one line change to use that mechanism for this specially marked binary, once we'd detected the section.

jasonmolenda committed:
format

jasonmolenda committed:
Add an additional aspect to the API test

In addition to a memory Module which has the __TEXT,__lldb_no_nlist
section indicating that we do not read the symbol table for this
binary, I added a second library which is also a memory Module but
no section -- test that lldb did read the symbol table for this
binary.

JDevlieghere commented on lldb/test/API/macosx/no-nlist-memory-module/Makefile:
```diff
@@ -0,0 +1,19 @@
+	$(LD) -dynamiclib -o libhas-nlists.dylib has-nlists.o  -install_name "@executable_path/libhas-nlists.dylib"
+
+clean::
+	rm -rf has-nlists.o no-nlists.o no-nlist-sect.o main.o main.s a.out a.out.dSYM libno-nlists.dylib libno-nlists.dylib.dSYM libhas-nlists.dylib libhas-nlists.dylib.dSYM
```
Would it be easier to `rm -rf *.o *.dSYM *.dylib`?

jasonmolenda replied:
yeah tbh I don't think we ever even run the clean targets these days, I just stuck this in there out of habit.  good suggestion.

JDevlieghere commented on lldb/test/API/macosx/no-nlist-memory-module/TestNoNlistsDylib.py:
```diff
@@ -0,0 +1,81 @@
+        # modules from the Debugger so we don't hold on to a reference
+        # of the on-disk binary.
+
+        # If we haven't succeeded after ten attemps of attaching and
```
I don't really understand the need for this loop. Why would this fail the first time and succeed the second time? Is this trying to work around the debugger running in asynchronous mode or something?

JDevlieghere replied:
I figured out the answer when reading the test below: because the inferior is removing the dylib and there's no synchronization with the test. 

But that leads to another question: why are you not doing this from Python in the test itself? Seems like that would simplify things a lot (the test could just sleep forever)  and allow you to do the attach when you know the program is in the expected state?

jasonmolenda replied:
I tried doing in python first -- but I ended up removing the dylib before it loaded in the process, and it crashed starting up.  I'd have to add a sleep or some other sync to guarantee that the inferior had started and loaded the dylib before I remove it.

JDevlieghere replied:
Okay, I think I understand the challenge: you can't use lldb to synchronize (e.g. by waiting on a `pause()`) because by attaching, you're already going to be parsing the library. We have a few other tests that have this issue, and they use a file to synchronize through `lldbutil.wait_for_file_on_target`. 

```
        # Use a file as a synchronization point between test and inferior.
        pid_file_path = lldbutil.append_to_process_working_directory(
            self, "pid_file_%d" % (int(time.time()))
        )
        self.addTearDownHook(
            lambda: self.run_platform_command("rm %s" % (pid_file_path))
        )

        popen = self.spawnSubprocess(exe, [pid_file_path])

        pid = lldbutil.wait_for_file_on_target(self, pid_file_path)
```

I think we could use the same approach here.

jasonmolenda replied:
Good point.

jasonmolenda committed:
simplify makefile clean line

jasonmolenda committed:
Rewrite test case copying the sync scheme from TestAttachDenied.py

jasonmolenda committed:
sort includes to appease ci checker

jasonmolenda committed:
Remove the dylibs in the Python instead of the C file,
simpler now that we have the sync point when the
inferior app has started up & loaded the dylibs.

jasonmolenda commented:
I adopted the synchronization scheme that we do in other attach tests -- the python provides a filename that the inferior should create with its pid, we launch it and wait until that file exists (indicating that the inferior is done setting up) and then the python tests execute.

I changed the code which removes the two dylibs (so lldb is forced to read them out of memory) from the inferior main.c to the Python, where it is simpler.  I think we're good to go at this point.

JDevlieghere reviewed (approved):
LGTM. I think you could simplify the test even further by using C++ but on the other hand the C tests compile a little faster.

```
std::ofstream pid_file;
pid_file.open (pid_file_path);
pid_file << getpid(); 
pid_file.close();
```

jasonmolenda commented:
> LGTM. I think you could simplify the test even further by using C++ but on the other hand the C tests compile a little faster.
 
yeah I was going to rewrite it into slightly nicer C but then I realized there was more value in doing it exactly the same as the other test that I copied it from, so they could be both found by a simple search.

jasonmolenda commented:
FTR on the CI bot my API test did not make correctly.  I reverted the PR, rewrote the makefiles for the API test, relanded.

labath commented:
>  I create the specially named section by compiling an assembly file that puts a byte in the section which makes for a bit of a messy Makefile

You probably don't need an asm file for that. There are at least [two ways](https://godbolt.org/z/4sTrdrT49) to generate a random section from C:
```
asm(R"(.section ".text.foo"; .byte 0)");

void  __attribute__((section(".text.bar"))) f(){}
```
